pkg/querier/queryrange/results_cache.go

```diff
@@ -257,6 +266,44 @@ func (s resultsCache) shouldCacheResponse(ctx context.Context, r Response) bool
 	return true
 }
 
+func atModifierAfterEnd(r Request) bool {
+	query := r.GetQuery()
+	if !strings.Contains(query, "@") {
+		return false
+	}
+	expr, err := parser.ParseExpr(query)
+	if err != nil {
+		// We are being pessimistic in such cases.
+		return true
+	}
+
+	end := r.GetEnd()
+	atModAfterEnd := false
+	parser.Inspect(expr, func(n parser.Node, _ []parser.Node) error {
+		switch e := n.(type) {
+		case *parser.VectorSelector:
+			if e.Timestamp != nil && *e.Timestamp > end {
+				atModAfterEnd = true
+				return errors.New("at modifier after end")
```

Comment:
Nit: Since we're going to run this method for every single query, perhaps we can just create this sentinel error value once, and not on every invocation.